modules/DGM/PSO.cpp

```diff
@@ -0,0 +1,215 @@
+}
+
+void DirectGraphicalModels::PSO::setInitParams(const vec_float_t &vParams) {
+    DGM_ASSERT_MSG(m_vParams.size() == vParams.size(),
+                   "The size of the argument (%zu) does not correspond to the number of parameters (%zu)",
+                   vParams.size(), m_nParams);
+
+    for (size_t p = 0; p < vParams.size(); p++) {
+        const float &param = vParams[p];
+        if (param > m_vMax[p]) {
+            DGM_WARNING("Argument[%zu]=%.2f exceeds the upper boundary %.2f and will not be set",
+                        p, param, m_vMax[p]);
+            continue;
+        }
+        if (param < m_vMin[p]) {
+            DGM_WARNING("Argument[%zu]=%.2f exceeds the lower boundary %.2f and will not be set",
+                        p, param, m_vMin[p]);
+            continue;
+        }
+        m_vParams[p] = param;
+    }
+}
+
+void DirectGraphicalModels::PSO::setMinParams(const vec_float_t &vMinParam) {
+    DGM_ASSERT_MSG(m_vMin.size() == vMinParam.size(),
+                   "The size of the argument (%zu) does not correspond to the number of parameters (%zu)",
+                   vMinParam.size(), m_nParams);
+
+    for (size_t p = 0; p < vMinParam.size(); p++) {
+        const float &minParam = vMinParam[p];
+        if (minParam > m_vParams[p])
+            DGM_WARNING(
+                    "Argument[%zu]=%.2f contradicts the parameter value %.2f and will not be set",
+                    p, minParam, m_vParams[p]);
+        else m_vMin[p] = minParam;
+    }
+}
+
+void DirectGraphicalModels::PSO::setMaxParams(const vec_float_t &vMaxParam) {
+    DGM_ASSERT_MSG(m_vMax.size() == vMaxParam.size(),
+                   "The size of the argument (%zu) does not correspond to the number of parameters (%zu)",
+                   vMaxParam.size(), m_nParams);
+
+    for (size_t p = 0; p < vMaxParam.size(); p++) {
+        const float &maxParam = vMaxParam[p];
+        if (maxParam < m_vParams[p])
+            DGM_WARNING(
+                    "Argument[%zu]=%.2f contradicts the parameter value %.2f and will not be set",
+                    p, maxParam, m_vParams[p]);
+        else m_vMax[p] = maxParam;
+    }
+}
+
+vec_float_t DirectGraphicalModels::PSO::getParams(float (*objectiveFunct)(vec_float_t)) {
+    if (isThreadsEnabled) {
+        std::vector<std::thread> threads_v;
+        for (size_t i = 0; i < NUMBER_BOIDS; i++) {
+            threads_v.emplace_back(&DirectGraphicalModels::PSO::runPSO_withThreads,
+                    this, objectiveFunct, i);
+        }
+        std::for_each(threads_v.begin(), threads_v.end(), [](std::thread& th) {
+            th.join();
+        });
+
+        return gBest;
+    } else {
+        runPSO(objectiveFunct);
+        return gBest;
+    }
+}
+
+void DirectGraphicalModels::PSO::runPSO(float (*objectiveFunct)(vec_float_t)) {
+    size_t it = 0;
+    while (it < MAX_NR_ITERATIONS) {
+        for (auto i = 0; i < NUMBER_BOIDS; i++) {
```

Comment:
I think, it makes sense to use size of the vector here instead of the constant NUMBER_BOIDS...
Then it would be possible to use a short notation:
`for(auto &boid: m_vBoids)
   ...`